src/async_fuse/memfs/kv_engine/etcd_impl.rs

```diff
@@ -1,18 +1,246 @@
+            .field("lease_id", &self.lease_id)
+            .field("ttl", &self.ttl)
+            .finish_non_exhaustive()
+    }
+}
+
+impl SessionInner {
+    /// Get the closed flag
+    fn is_closed(&self) -> bool {
+        self.closed.load(Ordering::Acquire)
+    }
+
+    /// Set close flag
+    fn close(&self) {
+        self.closed.store(true, Ordering::Release);
+    }
+
+    /// Reopen close flag
+    fn reopen(&self) {
+        self.closed.store(false, Ordering::Release);
+    }
+
+    /// Keep alive the lease
+    #[allow(clippy::pattern_type_mismatch)] // Raised by `tokio::select`
+    pub async fn keepalive(&self, mut close_rx: mpsc::Receiver<()>, token: CancellationToken) {
+        let mut interval = tokio::time::interval(Duration::from_secs(self.ttl.div_ceil(3)));
+
+        if self.is_closed() {
+            error!("lease keep alive stream closed by is_closed flag");
+            return;
+        }
+
+        // Try to clone a client, if the keeper failed, try to reconnect
+        let mut client = self.client.clone();
+        let lease_id = self.lease_id;
+        let (mut keeper, mut lease_keep_alive_stream) =
+            match client.lease_keep_alive(lease_id).await {
+                Ok((keeper, lease_keep_alive_stream)) => (keeper, lease_keep_alive_stream),
+                Err(e) => {
+                    error!("failed to keep alive lease, error={e:?}");
+
+                    return;
+                }
+            };
+
+        // Set the is_closed flag
+        self.reopen();
+
+        // Used to revoke client
+        let mut client = self.client.clone();
+        // Start to keep alive the lease
+        loop {
+            if self.is_closed() {
+                error!("lease keep alive stream closed by is_closed flag");
+                return;
+            }
+
+            tokio::select! {
+                _ = interval.tick() => {
+                    // Try to send a keep alive request
+                    match keeper.keep_alive().await {
+                        Ok(o) => {
+                            info!("keep alive lease, lease_id={lease_id}, {o:?}");
+                        }
+                        Err(e) => {
+                            error!("failed to keep alive lease, error={e:?}");
+                            break;
+                        }
+                    }
+
+                    // Try to parse
+                    match lease_keep_alive_stream.message().await {
+                        Ok(Some(lease_alive_response)) => {
+                            if lease_alive_response.ttl() == 0 {
+                                error!("lease keep alive stream closed");
+                                self.close();
+                                return;
+                            }
+                            continue
+                        }
+                        Ok(None) => {
+                            error!("lease keep alive stream closed");
```

Comment:
Should confirm with the etcd client code, why there're two case to indicate stream close ?  
 `Ok(None)` and `lease_alive_response.ttl == 0 `

Reply:
Here are the comments from `tonic-0.9.2/src/codec/decode.rs`, `None` means that 
 the stream was closed by sender, `ttl is 0` means that the lease was outdated or revoked.
```rs
    /// Fetch the next message from this stream.
    ///
    /// # Return value
    ///
    /// - `Result::Err(val)` means a gRPC error was sent by the sender instead
    /// of a valid response message. Refer to [`Status::code`] and
    /// [`Status::message`] to examine possible error causes.
    ///
    /// - `Result::Ok(None)` means the stream was closed by the sender and no
    /// more messages will be delivered. Further attempts to call
    /// [`Streaming::message`] will result in the same return value.
    ///
    /// - `Result::Ok(Some(val))` means the sender streamed a valid response
    /// message `val`.
```

Reply:
Reference link is https://github.com/etcd-io/cetcd/blob/d3e3bdbcfd9609e0a5edd5315573de564ed723cb/vendor/github.com/coreos/etcd/clientv3/lease.go#L208. @frostyplanet